include/nbl/video/IGPUCommandPool.h

```diff
@@ -13,29 +14,386 @@ namespace nbl::video
+        static void* operator new[](std::size_t size) = delete;
+        // static void* operator new(std::size_t size, std::align_val_t al) = delete;
+        static void* operator new[](std::size_t size, std::align_val_t al) = delete;
+
+        // static void operator delete  (void* ptr) = delete;
+        static void operator delete[](void* ptr) = delete;
+        static void operator delete  (void* ptr, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::align_val_t al) = delete;
+        static void operator delete  (void* ptr, std::size_t sz) = delete;
+        static void operator delete[](void* ptr, std::size_t sz) = delete;
+        static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+        inline uint32_t getSize() const { return m_size; }
+
+    protected:
+        ICommand(uint32_t size) : m_size(size) {}
+
+    private:
+        friend CCommandSegment;
+
+        const uint32_t m_size;
+    };
+
+    template <class CRTP>
+    class NBL_FORCE_EBO IFixedSizeCommand : public IGPUCommandPool::ICommand
+    {
     public:
-        enum E_CREATE_FLAGS : uint32_t
+        template <typename... Args>
+        static uint32_t calc_size(const Args&...)
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            return core::alignUp(sizeof(CRTP), alignof(CRTP));
+        }
+
+    protected:
+        IFixedSizeCommand() : ICommand(calc_size()) {}
+    };
+
+    class alignas(COMMAND_SEGMENT_ALIGNMENT) CCommandSegment
+    {
+    private:
+        struct header_t
+        {
+            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+            CCommandSegment* m_next;
+        } header;
+
+    public:
+        static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - core::roundUp(sizeof(header_t), alignof(ICommand));
+
+        struct Iterator
+        {
+            ICommand* m_cmd = nullptr;
+            // uint32_t m_size = 0;
+            CCommandSegment* m_segment = nullptr;
+
+            inline bool operator!=(const Iterator& other) const { return m_cmd != other.m_cmd; }
+            // TODO(achal): Return Iterator&?
+            // inline void operator++()
+            // {
+            //     m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size);
+            //     m_size = ;
+            // }
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        CCommandSegment()
+        {
+            header.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, STORAGE_SIZE);
```

Comment:
After this https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r863768889, I forgot to change the last param to `STORAGE_SIZE` (it was `COMMAND_SEGMENT_SIZE` before) which caused me a lot of headache. I also think this was the reason behind the weird behaviour with `CPipelineBarrierCmd` I mentioned in this comment https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#issuecomment-1221468912.